### ISSUE TITLE
Ollama Utilizing Only CPU Instead of GPU on MacBook Pro M1 Pro

### ISSUE THREAD

**Issue by vidvudsc** [jmorganca/ollama#1986]
Description
I've encountered an issue where Ollama, when running any llm is utilizing only the CPU instead of the GPU on my MacBook Pro with an M1 Pro chip. This results in less efficient model performance than expected.

Environment
MacBook Pro with M1 Pro chip
MacOS version: Sonoma 14.2.1 
Ollama version: 1.20

No specific error messages are observed.
All dependencies and drivers are up to date.
I would appreciate any guidance or updates regarding this issue. If there are any configurations or settings I might be missing, please let me know.

<img width="852" alt="Screenshot 2024-01-14 at 08 00 10" src="https://github.com/jmorganca/ollama/assets/77242455/ee3c0398-37e9-4473-af5b-a3b3253d1662">

PS. the image was taken when running dolphin-mixtral

Thanks!

**Comment by easp** [comment]
16GB isn't nearly enough to run dolphin-mixtral at any reasonable speed. The default download is 26GB in size. The computer will have to move more than 10GB of data from the SSD for every token generated. 

This isn't really practical when using the GPU (or at all, really) so Ollama falls back to CPU. Under these conditions the difference between using CPU and GPU is insignificant, anyway since most of the time is spent moving data from the SSD. 

Because it spends most of the time waiting for data transfer from the SSD, the CPU is largely idle.

The model data is memory mapped and so it's not accounted for in normal process memory. It should be accounted for in wired memory and/or file cache.

In short, your expectations are out of line with realities of what your computer is capable of and how resource use is accounted for.

As for what you can do... For reasonable performance, run models that fit within the memory that MacOS makes accessible to the GPU (66% of 16GB by default, which is about 10.5GB). That's not going to be enough for even a ~2-bit quantization of Mixtral.

**Comment by jmorganca** [comment]
Hi there, what @easp mentioned is a great overview of why it uses the CPU right now. At the moment Ollama won't partially use the GPU, it will fall back to CPU. That said look out for improvements to this in the future. For your setup smaller models should run quite fast on the GPU (e.g. `llama2`, `mistral`)

**Comment by jmorganca** [comment]
Also, thanks @easp !

**Comment by vidvudsc** [comment]
@jmorganca @easp Thanks for the help! Really appreciated it.

**Comment by mdl054** [comment]
> 16GB isn't nearly enough to run dolphin-mixtral at any reasonable speed. The default download is 26GB in size. The computer will have to move more than 10GB of data from the SSD for every token generated.
> 
> This isn't really practical when using the GPU (or at all, really) so Ollama falls back to CPU. Under these conditions the difference between using CPU and GPU is insignificant, anyway since most of the time is spent moving data from the SSD.
> 
> Because it spends most of the time waiting for data transfer from the SSD, the CPU is largely idle.
> 
> The model data is memory mapped and so it's not accounted for in normal process memory. It should be accounted for in wired memory and/or file cache.
> 
> In short, your expectations are out of line with realities of what your computer is capable of and how resource use is accounted for.
> 
> As for what you can do... For reasonable performance, run models that fit within the memory that MacOS makes accessible to the GPU (66% of 16GB by default, which is about 10.5GB). That's not going to be enough for even a ~2-bit quantization of Mixtral.

Sorry to hijack, does this mean having more RAM means you can load larger models or do you mean that 16GB is a hard limit due to the memory the GPU has available? So if we had a Mac with 96gb vs the 16gb for example.

**Comment by easp** [comment]
@mdl054 If you have more RAM you can load larger models and have them processed on the GPU. MacOS gives the GPU access to 2/3rds of system memory on Macs with 36GB or less and 3/4 on machines with 48GB or more. A 96GB Mac has 72 GB available to the GPU. Some of that will be needed beyond the model data itself.

There is a way to allocate more RAM to the GPU, but as of 0.1.22 Ollama doesn't take it into account.

**Comment by srbt1** [comment]
Thanks so much @easp : that answers my question too :)

**Comment by iamshreeram** [comment]
Hi @easp, I'm using ollama to run models on my old MacBook Pro with an Intel (i9 with 32GB RAM) and an AMD Radeon GPU (4GB).

Despite setting the environment variable OLLAMA_NUM_GPU to 999, the inference process is primarily using 60% of the CPU and not the GPU. 

Model I'm trying to run : starcoder2:3b (1.7 GB).

As the model size is low, How can I configure the inference to utilize the GPU fully without relying on the CPU at all?
